File: backend/app/routers/admin/clients.py
from fastapi import APIRouter, Depends, HTTPException
from app.database import get_database
from app.middleware.auth import require_admin
from app.schemas.user import CreateClientRequest, UserResponse
from app.models.user import new_user
from app.services.auth import hash_password, create_invite_token
from app.services.email import send_invite_email

router = APIRouter()


@router.get("")
async def list_clients(admin: dict = Depends(require_admin)):
    db = get_database()
    cursor = db.users.find({"role": "client"}).sort("created_at", -1)
    items = []
    async for user in cursor:
        items.append(UserResponse(
            id=str(user["_id"]),
            email=user["email"],
            role=user["role"],
            name=user.get("name", ""),
            phone=user.get("phone"),
            is_active=user["is_active"],
        ))
    return {"clients": items}


@router.post("", status_code=201)
async def create_client(body: CreateClientRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    existing = await db.users.find_one({"email": body.email})
    if existing:
        raise HTTPException(status_code=400, detail="User with this email already exists")
    invite_token = create_invite_token()
    user = new_user(
        email=body.email,
        password_hash=hash_password(invite_token),
        role="client",
        name=body.name,
    )
    user["phone"] = body.phone
    user["invite_token"] = invite_token
    result = await db.users.insert_one(user)
    send_invite_email(body.email, body.name, invite_token)
    return {"id": str(result.inserted_id), "message": "Client created and invite sent"}

File: src/routes/admin/clients.ts
import { Router, Response } from 'express';
import { getDb } from '../../database';
import { requireAdmin, AuthRequest } from '../../middleware/auth';
import { newUser } from '../../models/user';
import { hashPassword, createInviteToken } from '../../services/auth';
import { sendInviteEmail } from '../../services/email';

const router = Router();

router.get('/', requireAdmin, async (_req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const clients = await db.collection('users').find({ role: 'client' }).sort({ created_at: -1 }).toArray();
  res.json({ clients: clients.map((u) => ({ id: u._id.toString(), email: u.email, role: u.role, name: u.name ?? '', phone: u.phone ?? null, is_active: u.is_active })) });
});

router.post('/', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const { email, name, phone } = req.body;
  if (!email || !name) { res.status(400).json({ detail: 'Email and name required' }); return; }
  const db = await getDb();
  const existing = await db.collection('users').findOne({ email });
  if (existing) { res.status(400).json({ detail: 'User with this email already exists' }); return; }
  const inviteToken = createInviteToken();
  const user = newUser(email, hashPassword(inviteToken), 'client', name);
  (user as Record<string, unknown>).phone = phone ?? null;
  (user as Record<string, unknown>).invite_token = inviteToken;
  const result = await db.collection('users').insertOne(user);
  // Invite email intentionally not sent — client portal accounts not yet enabled
  res.status(201).json({ id: result.insertedId.toString(), message: 'Client created' });
});

export default router;
